pkg/cloudprovider/providers/gce/gce_util.go

```diff
@@ -120,12 +123,47 @@ func lastComponent(s string) string {
 // mapNodeNameToInstanceName maps a k8s NodeName to a GCE Instance Name
 // This is a simple string cast.
 func mapNodeNameToInstanceName(nodeName types.NodeName) string {
-	return string(nodeName)
+	return canonicalizeInstanceName(string(nodeName))
 }
 
 // mapInstanceToNodeName maps a GCE Instance to a k8s NodeName
-func mapInstanceToNodeName(instance *compute.Instance) types.NodeName {
-	return types.NodeName(instance.Name)
+func mapInstanceToNodeName(instance *compute.Instance) (types.NodeName, error) {
+	currentInstanceHostname, err := getInstanceHostname()
+	if err != nil {
+		return "", err
+	}
+	currentInstanceProjectId, err := getInstanceProjectViaMetadata()
+	if err != nil {
+		return "", err
+	}
+	return getNodeName(instance.Name, currentInstanceHostname, currentInstanceProjectId)
+}
+
+func getNodeName(instanceName string, instanceHostname string, instanceProjectId string) (types.NodeName, error) {
+	matched, err := regexp.MatchString(".*\\.c\\."+instanceProjectId+"\\.internal", instanceHostname)
+	if err != nil {
+		return "", err
+	}
+	if matched {
+		return types.NodeName(instanceName + ".c." + instanceProjectId + ".internal"), nil
```

Comment:
I'm not sure if this is a great idea. The naming scheme used here `instance.c.project-id.internal` is the DNS name of the instance and is not guaranteed to stay the same, and there could be multiple names pointing to the same instance...